draft-ietf-httpbis-resumable-upload.md

```diff
@@ -220,9 +220,14 @@ If the request completes successfully and the entire upload is complete, the ser
 
 If the request completes successfully but the entire upload is not yet complete, as indicated by an `Upload-Complete` field value of false in the request, the server MUST acknowledge it by responding with the `201 (Created)` status code and an `Upload-Complete` header value set to false.
 
-If the request includes an `Upload-Complete` field value set to true and a valid `Content-Length` header field, the client attempts to upload a fixed-length resource in one request. In this case, the upload's final size is the `Content-Length` field value and the server MUST record it to ensure its consistency.
+The request can indicate the upload's final size in two different ways. Both indicators may be present in the same request as long as they convey the same size. If the sizes are inconsistent, the server MUST reject the request by responding with a `400 (Bad Request)` status code.
 
-The server MAY enforce a maximum size of an upload resource. This limit MAY be equal to the upload's final size, if `Upload-Complete: ?1` and `Content-Length` are present in the upload creation request, or an arbitrary value. The limit's value or its existence MUST NOT change throughout the lifetime of the upload resource. The server MAY indicate such a limit to the client by including the `Upload-Limit` header field in the informational or final response to upload creation. If the client receives an `Upload-Limit` header field indicating that the maximum size is less than the amount of bytes it intends to upload to a resource, it SHOULD stop the current upload transfer immediately and cancel the upload ({{upload-cancellation}}).
+- If the request includes an `Upload-Complete` field value set to true and a valid `Content-Length` header field, the client attempts to upload a fixed-length resource in one request. In this case, the upload's final size is the `Content-Length` field value and the server MUST record it to ensure its consistency. The value cannot be used if the `Content-Length` header is missing (e.g. because of HTTP/1.1 chunked transfer coding, see {{Section 7.1 of RFC9112}}) or if the upload is split across multiple requests.
```

Comment:
It's not absolutely necessary to mention chunked transfer here, but I thought it's helpful to understand in which cases Content-Length may not be present. HTTP/1.1 is a bit different than H2 and H3 here since Content-Length must not be present when chunked transfer is used. In H2 and H3 this is not the case as they use a different method for message framing.

Do you think we should omit the HTTP/1.1 reference entirely? Or also mention H2 and H3 briefly?

Reply:
Personally I would like to kill h1 and don't want to remind people of h1 specific concepts. Maybe we can describe the concept in a version independent way? Something like "Content-Length is unknown upfront".

Reply:
I agree with Guoye here. Just drop the "e.g. because of HTTP/1.1...)